### PR TITLE
Extending unit test coverage for include/swift/AST/TensorFlow.h

### PR DESCRIPTION
**Pull request by vguerra**
Resolves [TF-46](https://bugs.swift.org/browse/TF-46).

Originally, TF-46 was about adding unit tests for [TFUtilities.h](https://github.com/apple/swift/blob/tensorflow/lib/SILOptimizer/Mandatory/TFUtilities.h) but after some discussion ( check out the ticket for details ) the goal changed to extend test coverage of [TensorFlow.h](https://github.com/apple/swift/blob/tensorflow/include/swift/AST/TensorFlow.h).

### CODE REVIEW

File: unittests/AST/TensorFlow.cpp [hunk 1]
```diff
@@ -77,9 +82,9 @@ class TypeContainsTensorFlowValueTest : public ::testing::Test {
 TEST_F(TypeContainsTensorFlowValueTest, ClassifiesCorrectly) {
```
Comment (bgogul) [inline unittests/AST/TensorFlow.cpp]: Given that you have expanded the test, why don't you change the name of the test fixture? Say TensorTypeUtilsTest?

File: unittests/AST/TensorFlow.cpp [hunk 2]
```diff
@@ -99,3 +104,78 @@ TEST_F(TypeContainsTensorFlowValueTest, WorksForRecursiveTypes) {
...
   EXPECT_FALSE(containsTensorFlowValue(recursiveType));
 }
+
+TEST_F(TypeContainsTensorFlowValueTest, TensorFlowType) {
+  EXPECT_TRUE(tf::isTensorHandle(tensorHandleType));
+  EXPECT_FALSE(tf::isTensorHandle(floatType));
+}
+
+TEST_F(TypeContainsTensorFlowValueTest, TensorHandleElementType) {
+  EXPECT_TRUE(tf::getTensorHandleElementType(floatType).isNull());
```
Comment (bgogul) [inline unittests/AST/TensorFlow.cpp]: Can you also add a test to check that the function extracts the correct type?

File: unittests/AST/TensorFlow.cpp [hunk 3]
```diff
@@ -99,3 +104,78 @@ TEST_F(TypeContainsTensorFlowValueTest, WorksForRecursiveTypes) {
   recursiveDecl->addMember(varDecl);
   EXPECT_FALSE(containsTensorFlowValue(recursiveType));
 }
+
+TEST_F(TypeContainsTensorFlowValueTest, TensorFlowType) {
```
Comment (bgogul) [inline unittests/AST/TensorFlow.cpp]: TensorFlowType -> IsTensorHandleWorks (or something along these lines)?

File: unittests/AST/TensorFlow.cpp [hunk 4]
```diff
@@ -99,3 +104,78 @@ TEST_F(TypeContainsTensorFlowValueTest, WorksForRecursiveTypes) {
...
+  EXPECT_TRUE(tf::isTensorHandle(tensorHandleType));
+  EXPECT_FALSE(tf::isTensorHandle(floatType));
+}
+
+TEST_F(TypeContainsTensorFlowValueTest, TensorHandleElementType) {
+  EXPECT_TRUE(tf::getTensorHandleElementType(floatType).isNull());
+  EXPECT_FALSE(tf::getTensorHandleElementType(tensorHandleType).isNull());
+}
+
+TEST_F(TypeContainsTensorFlowValueTest, ClassifyTensorFlowTypes) {
```
Comment (bgogul) [inline unittests/AST/TensorFlow.cpp]: Would be good to split this into separate tests.

File: unittests/AST/TensorFlow.cpp [hunk 5]
```diff
@@ -99,3 +104,78 @@ TEST_F(TypeContainsTensorFlowValueTest, WorksForRecursiveTypes) {
...
+            tf::classifyTensorFlowValue(variantHandleType));
+
+  EXPECT_EQ(tf::TFValueKind::Nope, tf::classifyTensorFlowValue(floatType));
+
+  auto *classDecl = testContext.makeNominal<ClassDecl>("SomeClass");
+  Type classType = classDecl->getDeclaredInterfaceType();
+  EXPECT_EQ(tf::TFValueKind::Nope, tf::classifyTensorFlowValue(classType));
+
+  // We test all functions defined around classifyTensorFlowValue
+  EXPECT_TRUE(tf::isTensorHandle(tensorHandleType));
```
Comment (bgogul) [inline unittests/AST/TensorFlow.cpp]: separate test?

File: unittests/AST/TensorFlow.cpp [hunk 6]
```diff
@@ -99,3 +104,78 @@ TEST_F(TypeContainsTensorFlowValueTest, WorksForRecursiveTypes) {
...
+  EXPECT_FALSE(tf::isTensorHandle(floatType));
+
+  EXPECT_TRUE(tf::isOpaqueHandle(resourceHandleType) &&
```
Comment (bgogul) [inline unittests/AST/TensorFlow.cpp]: separate test?

File: unittests/AST/TensorFlow.cpp [hunk 7]
```diff
@@ -99,3 +104,78 @@ TEST_F(TypeContainsTensorFlowValueTest, WorksForRecursiveTypes) {
...
+              tf::isOpaqueHandle(variantHandleType));
+  EXPECT_FALSE(tf::isOpaqueHandle(classType));
+
+  EXPECT_TRUE(tf::isTensorFlowValue(tensorHandleType) &&
```
Comment (bgogul) [inline unittests/AST/TensorFlow.cpp]: separate test?

File: unittests/AST/TensorFlow.cpp [hunk 8]
```diff
@@ -99,3 +104,78 @@ TEST_F(TypeContainsTensorFlowValueTest, WorksForRecursiveTypes) {
...
+  EXPECT_FALSE(tf::isTensorFlowValue(floatType));
+}
+
+TEST_F(TypeContainsTensorFlowValueTest, flattenTensorFlowValueAggregate) {
+  SmallVector<Type, 4> flattenedTensorFlowTypes;
+
+  auto *anyDecl = testContext.makeNominal<ClassDecl>("AnyHandle");
+  EXPECT_FALSE(tf::flattenTensorFlowValueAggregate(
+      anyDecl->getDeclaredInterfaceType(), flattenedTensorFlowTypes));
+  EXPECT_EQ(0ul, flattenedTensorFlowTypes.size());
```
Comment (bgogul) [inline unittests/AST/TensorFlow.cpp]: Could you check the contents of the returned types as well? I *think* you should be able to use matchers if you set up the actual and expected types to be canonical (See `Type->getCanonicalType`):
```
EXPECT_THAT(flattenedTensorFlowTypes, ::testing::ElementsAre(expected1, expected2, ...))
```

**Review by bgogul — CHANGES_REQUESTED** [review]
Thanks a lot for adding tests, @vguerra! 

Just left some comments on the PR.

File: unittests/AST/TensorFlow.cpp (hunk 1 above)
Comment (vguerra) [reply unittests/AST/TensorFlow.cpp]: yes, that makes sense. Will rename it.

File: unittests/AST/TensorFlow.cpp (hunk 2 above)
Comment (vguerra) [reply unittests/AST/TensorFlow.cpp]: sure. I am adding that check on my next commit.

File: unittests/AST/TensorFlow.cpp (hunk 3 above)
Comment (vguerra) [reply unittests/AST/TensorFlow.cpp]: I gave more expressive name to this and other tests in the file as well. Renaming coming in my next commit.

File: unittests/AST/TensorFlow.cpp (hunk 4 above)
Comment (vguerra) [reply unittests/AST/TensorFlow.cpp]: I did separate them on next commit.

File: unittests/AST/TensorFlow.cpp (hunk 5 above)
Comment (vguerra) [reply unittests/AST/TensorFlow.cpp]: same here.

File: unittests/AST/TensorFlow.cpp (hunk 6 above)
Comment (vguerra) [reply unittests/AST/TensorFlow.cpp]: same here.

File: unittests/AST/TensorFlow.cpp (hunk 7 above)
Comment (vguerra) [reply unittests/AST/TensorFlow.cpp]: same here.

File: unittests/AST/TensorFlow.cpp (hunk 8 above)
Comment (vguerra) [reply unittests/AST/TensorFlow.cpp]: Indeed it makes sense to check the contents of the vector, thank you for the pointers on how to do this.

Given that direct comparison is not possible on `Type` objects, I had to write my own `Matcher` to achieve this. So instead, I am doing something like:
```
EXPECT_THAT(flattenedTensorFlowTypes, ::testing::Each(CUSTOM_MATCHER))
```
Check out the next commit the details on my next commit pls :).

**Comment by vguerra** [comment]
Thank you for your remarks @bgogul .. I am pushing a new commit addressing all of them.

File: unittests/AST/TensorFlow.cpp [hunk 9]
```diff
@@ -72,17 +81,28 @@ class TypeContainsTensorFlowValueTest : public ::testing::Test {
...
+
+  Type createHandleType(StringRef typeName) {
+    auto *resourceHandleDecl = testContext.makeNominal<ClassDecl>(typeName);
+    return resourceHandleDecl->getDeclaredInterfaceType();
+  }
 };
 
-TEST_F(TypeContainsTensorFlowValueTest, ClassifiesCorrectly) {
+// Given that Type objects are not directly comparable we define our own Matcher
+MATCHER_P(IsSameTypeAs, refType,
```
Comment (bgogul) [inline unittests/AST/TensorFlow.cpp]: nice! :)

**Review by bgogul — APPROVED** [review]
Thanks!

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Comment by bgogul** [comment]
@swift-ci please clean test tensorflow

**Pull request merged**